language-plutus-core/src/Language/PlutusCore/Clone.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module Language.PlutusCore.Clone ( cloneType
+                                 ) where
+
+import           Control.Monad.State.Class
+import           Control.Monad.Trans.State hiding (get, modify)
+import qualified Data.IntMap               as IM
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Type
+import           PlutusPrelude
+
+type CloneSt = IM.IntMap (TyNameWithKind ())
+
+type CloneM = StateT CloneSt
+
+cloneEnvAssign :: MonadState CloneSt m
+               => Unique
+               -> TyNameWithKind ()
+               -> m ()
+cloneEnvAssign (Unique i) ty = modify (IM.insert i ty)
+
+cloneType :: MonadQuote m => Type TyNameWithKind a -> m (Type TyNameWithKind ())
+cloneType = flip evalStateT mempty . cloneTypeM
+
+cloneTypeM :: (MonadQuote m) => Type TyNameWithKind a -> CloneM m (Type TyNameWithKind ())
+cloneTypeM (TyFix _ (TyNameWithKind (TyName (Name l' n u))) ty) = do
+    u' <- liftQuote freshUnique
+    let tn = void $ TyNameWithKind (TyName (Name l' n u')) -- TODO use lenses
+    cloneEnvAssign u tn
+    TyFix () tn <$> cloneTypeM ty
+cloneTypeM (TyForall _ (TyNameWithKind (TyName (Name l' n u))) k ty) = do
+    u' <- liftQuote freshUnique
+    let tn = void $ TyNameWithKind (TyName (Name l' n u'))
+    cloneEnvAssign u tn
+    TyForall () tn (void k) <$> cloneTypeM ty
+cloneTypeM (TyLam _ (TyNameWithKind (TyName (Name l' n u))) k ty) = do
+    u' <- liftQuote freshUnique
+    let tn = void $ TyNameWithKind (TyName (Name l' n u'))
+    cloneEnvAssign u tn
+    TyLam () tn (void k) <$> cloneTypeM ty
```

Comment:
A lot of repetition in the three clauses above.

Reply:
Still the case that there is a lot of repetition. Probably there is an abstraction behind it.